contracts/smart-account/modules/SessionKeyManagers/SessionKeyManagerHybrid.sol

```diff
@@ -0,0 +1,806 @@
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {BaseAuthorizationModule} from "../BaseAuthorizationModule.sol";
+import {IAuthorizationModule} from "../../interfaces/IAuthorizationModule.sol";
+import {ISmartAccount} from "../../interfaces/ISmartAccount.sol";
+
+/**
+ * @title Session Key Manager module for Biconomy Modular Smart Accounts.
+ * @dev Stores the Session Information explicity in the storage, instead of maintainting
+ *      a merkle tree.
+ *      This reduces the amount of calldata required to validate a session key, making it cheaper on
+ *      L2s.
+ *      Allows for a session to be enabled explicity, or being batched with the first usage of said session
+ * @author Ankur Dubey - <ankur@biconomy.io>
+ * @author Fil Makarov - <filipp.makarov@biconomy.io>
+ */
+contract SessionKeyManagerHybrid is
+    BaseAuthorizationModule,
+    ISessionKeyManagerModuleHybrid
+{
+    using UserOperationLib for UserOperation;
+
+    uint256 private constant MODULE_SIGNATURE_OFFSET = 96;
+
+    mapping(bytes32 _sessionDataDigest => mapping(address _sa => SessionData data))
+        internal _enabledSessionsData;
+
+    /// @inheritdoc IAuthorizationModule
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external virtual override returns (uint256 rv) {
+        if (_isBatchExecuteCall(userOp)) {
+            rv = _validateUserOpBatchExecute(userOp, userOpHash);
+        } else {
+            rv = _validateUserOpSingleExecute(userOp, userOpHash);
+        }
+    }
+
+    /***************************** Single Call Handler ***********************************/
+
+    function _validateUserOpSingleExecute(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) internal returns (uint256 rv) {
+        /*
+         * Module Signature Layout
+         * Offset (in bytes)    | Length (in bytes) | Contents
+         * 0x0                  | 0x1               | 0x01 if sessionEnableTransaction, 0x00 otherwise
+         * 0x1                  | --                | Data depending on the above flag
+         */
+        bytes calldata moduleSignature = userOp
+            .signature[MODULE_SIGNATURE_OFFSET:];
+
+        if (_isSessionEnableTransaction(moduleSignature)) {
+            (
+                uint256 sessionKeyIndex,
+                uint48 validUntil,
+                uint48 validAfter,
+                address sessionValidationModule,
+                bytes calldata sessionKeyData,
+                bytes calldata sessionEnableData,
+                bytes calldata sessionEnableSignature,
+                bytes calldata sessionKeySignature
+            ) = _parseSessionEnableSignatureSingleCall(moduleSignature);
+
+            _verifySessionEnableDataSignature(
+                sessionEnableData,
+                sessionEnableSignature,
+                userOp.getSender()
```

Comment:
would this always be msg.sender?

Reply:
What do you mean by msg.sender here?

Reply:
like this will always be called by smart account

Reply:
Yes, the session enable data will always be signed by the Smart Account itself,.so signature verification has to happen again userOp.sender